code/modules/antagonists/traitor/datum_traitor.dm

```diff
@@ -23,43 +7,93 @@
+	///in the future, this should definitely be moved into a component that attaches to these datums
 	var/datum/contractor_hub/contractor_hub
 
+	///reference to the uplink this traitor was given, if they were.
+	var/datum/component/uplink/uplink
+
+	///how many more objectives past the starting bunch need to get finished before escape objective is unlocked
+	var/additional_objectives_before_escape = 4
+	///if new objectives won't be added because a previously finished objective got unfinished, lazylist reference to uncompleted objectives
+	var/list/new_objectives_blocked
+	///when new objectives are blocked, the objectives to add pile up here
+	var/objectives_to_add = 0
+	///the final objective the traitor has to accomplish, be it escaping, hijacking, or just martyrdom.
+	var/datum/objective/ending_objective
+
 /datum/antagonist/traitor/on_gain()
 	owner.special_role = job_rank
 	if(give_objectives)
 		forge_traitor_objectives()
+	//will be given later
+	//forge_ending_objective()
+	for(var/datum/objective/smart/objective in objectives)
+		RegisterSignal(objective, COMSIG_SMART_OBJECTIVE_ACHIEVED, .proc/objective_done)
+		RegisterSignal(objective, COMSIG_SMART_OBJECTIVE_UNACHIEVED, .proc/objective_undone)
+
+
+	var/faction = prob(75) ? FACTION_SYNDICATE : FACTION_NANOTRASEN
+	pick_employer(faction)
+
+	traitor_flavor = strings(TRAITOR_FLAVOR_FILE, employer)
+
+	if(give_uplink)
+		owner.give_uplink(silent = TRUE, antag_datum = src, ui_name = traitor_flavor["uplink_name"], ui_theme = traitor_flavor["uplink_theme"])
 
-	equip()
+	uplink = owner.find_syndicate_uplink()
 
 	owner.current.playsound_local(get_turf(owner.current), 'sound/ambience/antag/tatoralert.ogg', 100, FALSE, pressure_affected = FALSE, use_reverb = FALSE)
 
 	return ..()
 
 /datum/antagonist/traitor/on_removal()
 	if(!silent && owner.current)
-		to_chat(owner.current,"<span class='userdanger'>You are no longer the [job_rank]!</span>")
+		to_chat(owner.current,"<span class='userdanger'>You are no longer the [name]!</span>")
 
 	owner.special_role = null
 
 	return ..()
 
-/// Simple helper proc that handles equipping the traitor if necessary.
-/datum/antagonist/traitor/proc/equip()
-	if(should_equip)
-		owner.equip_traitor(employer, silent, src)
+/datum/antagonist/traitor/proc/pick_employer(faction)
+	var/possible_employers = list()
+	switch(faction)
+		if(FACTION_SYNDICATE)
+			possible_employers = GLOB.syndicate_employers.Copy()
+			if(istype(ending_objective, /datum/objective/hijack))
+				possible_employers -= GLOB.normal_employers
+			else //escape or martyrdom
+				possible_employers -= GLOB.hijack_employers
+		if(FACTION_NANOTRASEN)
+			possible_employers = GLOB.nanotrasen_employers.Copy()
+			if(istype(ending_objective, /datum/objective/hijack))
+				possible_employers -= GLOB.normal_employers
+			else //escape or martyrdom
+				possible_employers -= GLOB.hijack_employers
+	employer = pick(possible_employers)
```

Comment:
Change this to just return the `pick` result, and do `employer = pick_employer()`. Reducing side-effects within functions makes them easier to work with.